src/linear_algebra/exterior_algebra.lean

```diff
@@ -0,0 +1,145 @@
+/-
+Copyright (c) 2020 Adam Topaz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Zhangir Azerbayev, Adam Topaz, Eric Wieser.
+-/
+
+import algebra.ring_quot
+import linear_algebra.tensor_algebra
+import group_theory.perm.sign
+
+/-!
+# Exterior Algebras
+
+We construct the exterior algebra of a semimodule `M` over a commutative semiring `R`.
+
+## Notation
+
+The exterior algebra of the `R`-semimodule `M` is denoted as `exterior_algebra R M`.
+It is endowed with the structure of an `R`-algebra.
+
+Given a linear morphism `f : M → A` from a semimodule `M` to another `R`-algebra `A`, such that
+`cond : ∀ m : M, f m * f m = 0`, there is a (unique) lift of `f` to an `R`-algebra morphism,
+which is denoted `exterior_algebra.lift R f cond`.
+
+The canonical linear map `M → exterior_algebra R M` is denoted `exterior_algebra.ι R`.
+
+## Theorems
+
+The main theorems proved ensure that `exterior_algebra R M` satisfies the universal property
+of the exterior algebra.
+1. `ι_comp_lift` is  fact that the composition of `ι R` with `lift R f cond` agrees with `f`.
+2. `lift_unique` ensures the uniqueness of `lift R f cond` with respect to 1.
+
+## Implementation details
+
+The exterior algebra of `M` is constructed as a quotient of the tensor algebra, as follows.
+1. We define a relation `exterior_algebra.rel R M` on `tensor_algebra R M`.
+   This is the smallest relation which identifies squares of elements of `M` with `0`.
+2. The exterior algebra is the quotient of the tensor algebra by this relation.
+
+-/
+
+variables (R : Type*) [comm_semiring R]
+variables (M : Type*) [add_comm_monoid M] [semimodule R M]
+
+namespace exterior_algebra
+open tensor_algebra
+
+/--
+An inductively defined relation on `tensor_algebra R M` used to define the exterior algebra.
+-/
+inductive rel : tensor_algebra R M → tensor_algebra R M → Prop
+| of (m : M) : rel ((ι R m) * (ι R m)) 0
+
+end exterior_algebra
+
+/--
+The exterior algebra of an `R`-semimodule `M`.
+-/
+@[derive [inhabited, semiring, algebra R]]
+def exterior_algebra := ring_quot (exterior_algebra.rel R M)
+
+namespace exterior_algebra
+
+variables {M}
+
+/--
+The canonical linear map `M →ₗ[R] exterior_algebra R M`.
+-/
+def ι : M →ₗ[R] exterior_algebra R M :=
+  (ring_quot.mk_alg_hom R _).to_linear_map.comp (tensor_algebra.ι R)
+
+
+variables {R}
+
+/-- As well as being linear, `ι m` squares to zero -/
+@[simp]
+theorem ι_square_zero (m : M) : (ι R m) * (ι R m) = 0 :=
+begin
+  dsimp [ι],
+  rw [←alg_hom.map_mul, ←alg_hom.map_zero _],
+  exact ring_quot.mk_alg_hom_rel R (rel.of m),
+end
+
+variables (R)
+
+/--
+Given a linear map `f : M →ₗ[R] A` into an `R`-algebra `A`, which satisfies the condition:
+`cond : ∀ m : M, f m * f m = 0`, this is the canonical lift of `f` to a morphism of `R`-algebras
+from `exterior_algebra R M` to `A`.
+-/
+def lift {A : Type*} [semiring A] [algebra R A] (f : M →ₗ[R] A) (cond : ∀ m, f m * f m = 0) :
```

Comment:
I agree that `f` should be explicit here, to match e.g. `ideal.quotient.lift`. But could you move some of these arguments into a `variables` statement?

Reply:
This code is written this way to match

https://github.com/leanprover-community/mathlib/blob/da66bb81bf0466335bae82077f0c335dfe53aeb3/src/algebra/free_algebra.lean#L185-L189

Reply:
Then that file should also use more `variables` declarations :)

Reply:
Can that be a follow-up PR, or would you like me to clean up both in this one?

Reply:
Variables in `free_algebra` are fixed in #4338

Reply:
Thank you so much for the cleanup PR! I pushed [a commit](https://github.com/leanprover-community/mathlib/commit/daf239dab30bae1ea8373ae38d5e8b1c7750ecdd) to a new branch, adding the `variables` statement. If you agree with it, could you add it to this branch?

Reply:
I'm not sure I agree with "hiding" `cond` and `f` like that, especially given the need for `include` and `omit` - but I'm definitely in favor of moving all the type variables and classes.

Reply:
That's very reasonable :) How about 93cad361?

Reply:
Looks good to me